sklearn/preprocessing/_data.py

```diff
@@ -765,28 +783,103 @@ def partial_fit(self, X, y=None):
                 raise ValueError(
                     "Cannot center sparse matrices: pass `with_mean=False` "
                     "instead. See docstring for motivation and alternatives.")
-
             sparse_constructor = (sparse.csr_matrix
                                   if X.format == 'csr' else sparse.csc_matrix)
-            counts_nan = sparse_constructor(
+
+            if sample_weight is not None:
+                counts_nan = sparse_constructor(
+                    (np.isnan(X.data), X.indices, X.indptr),
+                    shape=X.shape,
+                    dtype=sample_weight.dtype)
+                if counts_nan.format == 'csr':
+                    counts_nan.data *= sample_weight.repeat(
+                        np.diff(counts_nan.indptr))
+                elif counts_nan.format == 'csc':
+                    counts_nan.data *= sample_weight[counts_nan.row]
+                counts_nan = counts_nan.sum(axis=0).A.ravel()
```

Comment:
I think this could be done directly with the multiply method of the sparse matrix + broadcasting:

```python

                nan_values = sparse_constructor(
                    (np.isnan(X.data), X.indices, X.indptr),
                    shape=X.shape,
                    dtype=sample_weight.dtype)
                 nan_weights = nan_values.multiply(sample_wights[:, np.newaxis])
                 counts_nan = nan_weights.sum(axis=0).A.ravel()
```

and maybe rename `counts_nan` to `nan_weights`